pkg/driver/nydus/packer/packer.go

```diff
@@ -0,0 +1,304 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packer
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/containerd/containerd/content"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/goharbor/acceleration-service/pkg/driver/nydus/builder"
+	"github.com/goharbor/acceleration-service/pkg/driver/nydus/utils"
+)
+
+type Descriptor struct {
+	Blob      *ocispec.Descriptor
+	Bootstrap ocispec.Descriptor
+}
+
+// Packer implements the build workflow of nydus image, as well as
+// the export and import of build cache.
+type Packer struct {
+	parentWorkDir string
+	builderPath   string
+	sg            singleflight.Group
+}
+
+func New(parentWorkDir, builderPath string) (*Packer, error) {
+	return &Packer{
+		parentWorkDir: parentWorkDir,
+		builderPath:   builderPath,
+	}, nil
+}
+
+func (p *Packer) prepareWorkdir() (string, func() error, error) {
+	workDir, err := ioutil.TempDir(p.parentWorkDir, "nydus-build-")
+	if err != nil {
+		return "", nil, errors.Wrapf(err, "create work dir")
+	}
+
+	// Create a directory to store nydus blob file for every layer.
+	blobDirPath := path.Join(workDir, "blobs")
+	if err := os.MkdirAll(blobDirPath, 0755); err != nil {
+		return "", nil, errors.Wrapf(err, "create blob dir %s", blobDirPath)
+	}
+
+	// Create a directory to store nydus bootstrap file for every layer.
+	bootstrapDirPath := path.Join(workDir, "bootstraps")
+	if err := os.MkdirAll(bootstrapDirPath, 0755); err != nil {
+		return "", nil, errors.Wrapf(err, "create bootstrap dir %s", bootstrapDirPath)
+	}
+
+	cleanup := func() error {
+		return os.RemoveAll(workDir)
+	}
+
+	return workDir, cleanup, nil
+}
+
+func (p *Packer) diffBuild(ctx context.Context, workDir string, layers []*BuildLayer, diffSkip *int) (*builder.Output, error) {
+	diffPaths := []string{}
+	diffHintPaths := []string{}
+
+	bootstrapDir := path.Join(workDir, "bootstraps")
+	blobDir := path.Join(workDir, "blobs")
+	outputJSONPath := path.Join(workDir, "output.json")
+
+	for _, layer := range layers {
+		diffPaths = append(diffPaths, layer.diffPath)
+		diffHintPaths = append(diffHintPaths, layer.diffHintPath)
+	}
+
+	var parentBootstrapPath *string
+	if diffSkip != nil {
+		// Found nydus bootstrap cache, unpack targz and use it as parent bootstrap.
+		_parentBootstrapPath := path.Join(bootstrapDir, "parent-bootstrap")
+		parentBootstrapPath = &_parentBootstrapPath
+		layer := layers[*diffSkip]
+		cachedBootstrap, _ := layer.GetCache(ctx, CompressionTypeBootstrap)
```

Comment:
error shouldn't be ignored or should have a comment at least.

Reply:
Added a comment.